File: core-js/src/attest/httpSign.ts
import { randomBytes } from "node:crypto";
import { base64UrlEncode, canonicalize, sha256, sigInput, signEd25519, type AgentIdentity } from "openglass-sdk";

export class OpenGlassApiError extends Error {
  constructor(
    public method: string,
    public path: string,
    public status: number,
    public body: unknown,
  ) {
    super(`${method} ${path} -> ${status}: ${JSON.stringify(body)}`);
    this.name = "OpenGlassApiError";
  }
}

export type FetchLike = typeof fetch;

/**
 * SPEC §4.1 signed-request headers, built from the crypto primitives `openglass-sdk`
 * already exports (Ed25519 signing, canonical JSON, sha256) — this file only adds the
 * thin fetch/header-building wrapper on top, which `openglass-sdk`'s own client doesn't
 * currently expose as a standalone function. `fetchImpl` is injectable so integrations
 * (and this package's own tests) can simulate the network without a real server.
 */
export async function signedRequest<T = unknown>(
  baseUrl: string,
  method: string,
  path: string,
  body: unknown,
  identity: AgentIdentity,
  fetchImpl: FetchLike = fetch,
): Promise<T> {
  const bodyStr = body !== undefined ? JSON.stringify(body) : "";
  const bodySha256 = Buffer.from(sha256(Buffer.from(bodyStr, "utf8"))).toString("hex");
  const timestamp = new Date().toISOString();
  const nonce = base64UrlEncode(randomBytes(16));
  const digest = sha256(Buffer.from(canonicalize({ method, path, timestamp, nonce, bodySha256 }), "utf8"));
  const sig = base64UrlEncode(signEd25519(sigInput("request", digest), identity.privateKey));

  const headers: Record<string, string> = {
    "og-key": identity.kid,
    "og-timestamp": timestamp,
    "og-nonce": nonce,
    "og-signature": sig,
  };
  if (identity.agentId) headers["og-agent"] = identity.agentId;
  if (body !== undefined) headers["content-type"] = "application/json";

  const res = await fetchImpl(`${baseUrl}${path}`, { method, headers, body: body !== undefined ? bodyStr : undefined });
  const json = await res.json().catch(() => null);
  if (!res.ok) throw new OpenGlassApiError(method, path, res.status, json);
  return json as T;
}

File: core-py/src/openglass_core/attest/http_sign.py
"""SPEC §4.1 signed-request wrapper, ported from `openglass-sdk`'s own `http.py` (whose
`signed_request` isn't yet part of that package's public export surface) — the crypto
underneath (Ed25519 signing, canonical JSON, sha256) is entirely `openglass-sdk`'s;
this file only re-adds the thin httpx/header-building wrapper on top.
"""

from __future__ import annotations

import json
import secrets
from datetime import datetime, timezone
from typing import Any, Optional

import httpx
from openglass import AgentIdentity, base64url_encode, canonicalize, sha256, sig_input, sign_ed25519, to_hex


class OpenGlassApiError(Exception):
    def __init__(self, method: str, path: str, status: int, body: Any) -> None:
        super().__init__(f"{method} {path} -> {status}: {json.dumps(body)}")
        self.method = method
        self.path = path
        self.status = status
        self.body = body


def _now_iso() -> str:
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def signed_request(
    base_url: str,
    method: str,
    path: str,
    body: Any,
    identity: AgentIdentity,
    client: Optional[httpx.Client] = None,
) -> Any:
    """`client` is injectable so tests (and callers with their own connection pooling)
    can supply an `httpx.Client` bound to a `httpx.MockTransport` instead of a real
    server."""
    body_str = json.dumps(body, separators=(",", ":")) if body is not None else ""
    body_sha256 = to_hex(sha256(body_str.encode("utf-8")))
    timestamp = _now_iso()
    nonce = base64url_encode(secrets.token_bytes(16))
    digest = sha256(canonicalize({"method": method, "path": path, "timestamp": timestamp, "nonce": nonce, "bodySha256": body_sha256}).encode("utf-8"))
    sig = base64url_encode(sign_ed25519(sig_input("request", digest), identity.private_key))

    headers = {"og-key": identity.kid, "og-timestamp": timestamp, "og-nonce": nonce, "og-signature": sig}
    if identity.agent_id:
        headers["og-agent"] = identity.agent_id
    content = None
    if body is not None:
        headers["content-type"] = "application/json"
        content = body_str.encode("utf-8")

    owns_client = client is None
    c = client or httpx.Client()
    try:
        res = c.request(method, f"{base_url}{path}", headers=headers, content=content)
    finally:
        if owns_client:
            c.close()

    try:
        data = res.json()
    except ValueError:
        data = None
    if res.is_error:
        raise OpenGlassApiError(method, path, res.status_code, data)
    return data
